notifier: Extend capabilities of notifier to serve as event hub for all callback types

akloniex opened the pull request:
Notifier event handling is extended to handle more dynamic, nested callbacks. It also handles allocation of memory for said callbacks and unifies the interface across different event producers, as well as provides better scalability by allowing multiple callbacks being registered to single event.
The idea is to collect many different events into single event hub, for unification of interface and ease of use.

jajanusz commented on src/arch/host/CMakeLists.txt:
```diff
@@ -6,3 +6,5 @@ target_include_directories(sof_public_headers INTERFACE ${PROJECT_SOURCE_DIR}/sr
 # C & ASM flags
 target_compile_options(sof_options INTERFACE -g -O3 -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wimplicit-fallthrough=3)
+
+add_subdirectory(lib)
```
Please add empty newlines at end. I think checkpatch complains about it in C files but doesn't care about CMake.

akloniex replied:
done

jajanusz commented on src/include/sof/lib/notifier.h:
```diff
@@ -25,9 +25,21 @@ enum notify_id {
+	NOTIFIER_ID_SYNC_DMA_COPY,		/* struct dma_cb_data* */
+	NOTIFIER_ID_SYNC_DMA_IRQ,		/* struct dma_cb_data* */
+	NOTIFIER_ID_SYNC_ALL
+};
```
Maybe it'd be better to add `NOTIFIER_ID_SYNC_MAX = NOTIFIER_ID_SYNC_ALL`, because now this enum has 2 purposes - serves as 'broadcast' id and limit for loops

akloniex replied:
done

jajanusz commented on tools/logger/convert.c:
```diff
@@ -100,6 +100,7 @@ static const char * get_component_name(uint32_t component_id) {
 		CASE(SCHEDULE_LL);
 		CASE(ALH);
 		CASE(KEYWORD);
+		CASE(NOTIFIER);
```
Please also update get_trace_class in testbench

akloniex replied:
done

jajanusz commented on src/arch/host/lib/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+add_local_sources(sof notifier.c)
```
You can add newline if you are going to do more changes anyway. There is no need to retrigger Ci if everything else is ok.

akloniex replied:
done

jajanusz commented on src/include/sof/lib/notifier.h:
```diff
@@ -53,8 +65,25 @@ struct notify **arch_notify_get(void);
 
+void notifier_register_sync(void *receiver, void *caller,
+			    enum notify_sync_id type,
+		void (*cb)(void *arg, enum notify_sync_id type, void *data));
```
I guess this weird formatting is caused by the will to satisfy checkpatch...
I think in cases like that it's better to just indent all lines with 1 tab instead of having pyramids like this ;)

akloniex replied:
done

wenqingfu commented:
please check the consistent error of "aplay: set_params:1363: Unable to install hw params" on APL_UP2_NOCODEC in CI

tlauda commented on src/include/sof/lib/notifier.h:
```diff
@@ -29,6 +29,7 @@ enum notify_sync_id {
 	NOTIFIER_ID_SYNC_BUFFER_PRODUCE = 0,	/* struct buffer_cb_transact* */
 	NOTIFIER_ID_SYNC_BUFFER_CONSUME,	/* struct buffer_cb_transact* */
 	NOTIFIER_ID_SYNC_BUFFER_FREE,		/* struct buffer_cb_free* */
+	NOTIFIER_ID_SYNC_DMA_COPY,		/* struct dma_cb_data* */
```
Could you also refactor irq_callback added in https://github.com/thesofproject/sof/commit/4839742c35f37f54af4ce6169cf3e591db49b2ad to use new notifier API?

akloniex commented:
@tlauda I've refactored the original interface of notifier, introduced minor improvements at the same time. Then I've rebased my changes on top of that and made few changes regarding parts you've commented on.
Few changes are now in incorrect commits, as I still need to clean it up after rebases, will do that soon. Also DMA IRQ callbacks are not yet implemented with this new interface, will do that soon.
Please review these changes nonetheless, to verify that we're on the right track.

paulstelian97 commented:
> Also DMA IRQ callbacks are not yet implemented with this new interface, will do that soon.

Aren't DMA IRQ callbacks replaced by the DMA scheduler domain in purpose (@tlauda)? Or you mean actually triggering the scheduler itself?

tlauda commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,164 @@
-	list_item_del(&notifier->list);
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if ((!receiver || handle->receiver == receiver) &&
```
Let's add comment here what is your intention meaning what is your condition in searching for the right handles.

tlauda commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,164 @@
 }
 
-void notifier_event(struct notify_data *notify_data)
+void notifier_notify_remote(void)
```
I like this separation and avoiding unnecessary invalidations.

tlauda commented on src/lib/clk.c:
```diff
@@ -106,11 +104,12 @@ void clock_set_freq(int clock, uint32_t hz)
-	notify_data.message = CLOCK_NOTIFY_PRE;
-	notifier_event(&notify_data);
+	clk_pdata->clk_notify_data.message = CLOCK_NOTIFY_PRE;
+	notifier_event(NULL, notif_id, core_mask, &clk_pdata->clk_notify_data,
```
Let's add `clk_pdata->clk[clock]` as a caller to be consistent.

tlauda commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,164 @@
+				notify_data = &_notify_data[i];
+				notify_data->caller = caller;
+				notify_data->type = type;
+				notify_data->data = data;
```
I've just realized that some of the data here is passed from the local stack, so with asynchronous messages to other cores it still won't work correctly. Although it's not an issue right now and this problem has existed before your changes, we need to remember to have it fixed in the near future (we could allocate array with fixed max size and memcpy the data). For the sake of this PR such changes are not required.

akloniex replied:
@tlauda Notification payload evaluated on the same core might be located on stack, without causing any issues, and for multicore notifications, I've made sure that event payload is allocated within the 'caller', or is the caller itself.
So while there is the possibility to misuse multicore events with payload on stack, such instance doesn't exist at the moment. We might possibly add assert to verify that payload is outside of stack memory, to prohibit such future usages.

jajanusz replied:
IDK if it's ok with kernel guidelines, but maybe we should add comments like `// FXIME` or `// TODO` in cases like this, so when someone revisits that part of code he will see potential problem instantly.

paulstelian97 replied:
Some sort of `/* TODO: Check if attempting to pass on-stack data to a different core */ somewhere here. Although when talking about cross-core calls like with this notifier I'd expect callers to _be_ aware of such issues.

paulstelian97 commented on test/cmocka/src/notifier_mocks.c:
```diff
@@ -0,0 +1,102 @@
+	int i;
+
+	if (!_notify) {
+		_notify = rzalloc(0, 0, sizeof(*_notify));
```
Is this mock function used like that in other unit tests? Feels slightly weird to me to not just use malloc here.

paulstelian97 replied:
Requesting change on this one, looked more into it: The other tests and mocks actually indeed call `malloc` directly.

jajanusz commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,175 @@
+{
+	int i;
+
+	for (i = 0; i < NOTIFIER_ID_ALL; i++)
```
I think that's mistake. MAX should be used in cases like this (iterators) and ALL in case when you want to broadcast, they have the same value now but they don't have to.

paulstelian97 replied:
@jajanusz Shouldn't ALL then get some sort of reserved value like -1?

akloniex replied:
@paulstelian97 This comes from @jajanusz comment https://github.com/thesofproject/sof/pull/1950#discussion_r336417600 , but after @tlauda suggested to implement unsubscribe_all function, I believe that the 'ALL' enum value is obsolete. I'm going to remove it is a moment.

dbaluta commented:
@akloniex it looks like your change reveals a problem with IMX and host code. Please take care of the host code compilation error.

We will try to address the IMX one:

```
/home/sof/work/sof.git/src/lib/clk.c: In function 'clock_init':

/home/sof/work/sof.git/src/lib/clk.c:161:16: error: array subscript 1 is above array bounds of 'struct clk_data[1]' [-Werror=array-bounds]

  clk_pdata->clk[CLK_SSP].freq = ssp_freq[SSP_DEFAULT_IDX].freq;
```

CLK_SSP doesn't exist on IMX but now with your change the compiler complains about that :). Temporary fix will be for us to increase NUM_CLOCKS at 2, but we should really isolate accessing CLK_SSP into Intel specific code.

paulstelian97 commented:
For the host, can you check your CMakeLists? It seems to not find arch_notify_get even though you did add it in notifier_mocks.c.

Perhaps in test/cmocka/CMakeLists.txt around line 57, where universal_mocks is included?

paulstelian97 commented:
For i.MX, can you merge the patch I have made in a pull request to this branch? It would fix the build error but would in fact be a hack which we should have done a while ago. See https://github.com/akloniex/sof/pull/1

akloniex commented:
SOFCI TEST

akloniex commented on src/schedule/dma_multi_chan_domain.c:
```diff
@@ -250,9 +251,10 @@ static bool dma_multi_chan_domain_is_pending(struct ll_schedule_domain *domain,
-			/* execute callback if exists */
-			if (dmas[i].chan[j].irq_callback)
-				dmas[i].chan[j].irq_callback(&dmas[i].chan[j]);
+			notifier_event(&dmas[i].chan[j], NOTIFIER_ID_DMA_IRQ,
```
The performance is only slightly hindered, we're sacrificing a handful of clock cycles for ease of use, uniformity and scalability. The impact should be minimal.

paulstelian97 commented on src/arch/host/CMakeLists.txt:
```diff
@@ -6,3 +6,5 @@ target_include_directories(sof_public_headers INTERFACE ${PROJECT_SOURCE_DIR}/sr
 # C & ASM flags
 target_compile_options(sof_options INTERFACE -g -O3 -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wimplicit-fallthrough=3 -Wpointer-arith)
+
+add_subdirectory(lib)
```
Does this just make it work without including too much?

akloniex replied:
@paulstelian97 This only includes 'lib' subdirectory in current tree = 'src/arch/host/lib', which only contains notifier.c and CMakeLists.txt, the notifier implementation for host. I've added these in this PR, but during all the rebasing this slipped somewhere.
It only adds what is really needed.

lgirdwood commented on test/cmocka/src/notifier_mocks.c:
```diff
@@ -0,0 +1,102 @@
+	}
+}
+
+int notifier_register(void *receiver, void *caller, enum notify_id type,
```
Not following the API usage here, like why do we need receiver and caller.

akloniex replied:
All callbacks of given type are stored in separate lists.
Receiver is used:
- as arg passed back to the callback when event is triggered
- to match with receivers during unregister call, to only remove specified receiver from the list
Caller is used:
- to determine if triggered event should be propagated through given callback, so that filtering of events is done in notifier, not in receiver
- to match specific callback to remove via unregister

lgirdwood replied:
I'm aligned with the need for nesting and multiple reiceivers here, but the notifier is intended for notification of events. You are doing a messaging API which should have APIs that look more like Zephyr.

lgirdwood replied:
@akloniex This would be a lot better (as a notifying - not messaging API) if the producer and consumer had seperate registration APIs. e.g.
```
int notifier_consumer_register(void *consumer, type, callback);
int notifier_producer_register(void *producer, type);
```

akloniex replied:
@lgirdwood In my model, producers don't need to register at all. Maybe the confusion comes from lack of comments, or bad naming used.
Only consumers need to register, and the 'caller' parameter passed to registration is used to indicate to which producer we wish to register.
I could add some elaborate comments to these functions, if that would help to address the confusion.

lgirdwood replied:
ok, I'm thinking what happens if a reciever registers for a producer that does not exist (say not built in Kconfig option). We would continue atm, but would never be notified of the event.....I can live with this though.

The arg naming though should be more like:
```
notifier_register(type, callback, data);
```
Where callback is function and data is data passed to callback AND used for unregister. I am still not following why we need to pass the void * consumer and how it;s related to propogation.

akloniex replied:
@lgirdwood Think of it this way:
all events of given type, say buffer produce, are multiplexed via single `notify_id`;
there are many buffers in the topology and all contribute to the same event stream;
we might either propagate all these events to all consumers, forcing them to filter the events and react only to those that are relevant to them, or do the filtration in the notifier, based on the pointer to the producer, which was registered by consumer.

I went with the latter option, which saves at least a function call per consumer, but most often more than that, since we usually assign values to variables during declaration in the callback function, which is all wasted time if ultimately received event is not important for given consumer.

Regarding issue of registering to a producer that does not exist yet, that you mentioned:
there is not way of registering to a _specific_ producer when it does not exits, no reasonable deferred method or anything, so I believe we can skip the _specific_ part;
consumer might still register to any event type, with no specific producer in mind (passing `caller == NULL` to register function), effectively listening to all events that land under that `notify_id`, and then deduce if received event is relevant and take appropriate actions.

lgirdwood replied:
Thanks, I now see where you are going. This sounds like we are going to be broadcasting frquent events from multiple sources ? i.e. the "buffer produce" notification. It would be far better to use a more targeted notification here i.e. buffer X "produce" -> consumer Y, this needs a little bit more logic to setup but scales well.

The notification framework is more intended for low frequency broadcast events (e..g changing clock freq) and not suited for high frequency broacasting (i.e. the buffer produce event). This needs to be more targetted for high frequency events. 

Can you describe the event you are targetting here. Whats the use case.

akloniex replied:
I had 2 target applications in mind, both regarding buffer callbacks:
- probes
- echo reference
Both of which require support for multiple callbacks attached to one instance of buffer.

DMA callbacks got onto the train of changes mostly to unify usage of all callback types across the codebase.

Could you describe how would you like to implement the X->Y correspondence?
I believe that, for the most part, my solution gives that possibility, minimizing footprint and also giving the possibility to broadcast events. I'm kind of biased towards my solution since I've already written it, and can't think of a better way to implement this, but I'm open for fresh ideas on the subject.

lgirdwood replied:
So for both probes and echo reference we need to "messge by reference". i.e. "probe X is attatched to BUFFER1.0" by FDK, or "echo ref subscribes to BUFFER2.0 produce" set by topology config for ech ref.
The reference part can be a string based on module/comp name and module/comp ID (similar to CAVS FW).

lgirdwood replied:
@akloniex apologies for the delay. As discussed, the component/module messaging is different wrt to targetting and freqency with general system notifications. I think we need to define.
1) A set of component messages. i.e. "buffer produce", "buffer consume", "component stop", etc and add them as topology tuple tokens attributes for modules/components.
 2) Create an insertion macro so they can be added as attributes in topology components. This macro should also embed the source/target component ID.
3) Add light hooks into pipeline, buffer and components to send the messages. The messaging  mappings will be bound at pipeline creation time, i..e no need to search and map at runtime.

akloniex replied:
@lgirdwood The issue I have with your approach is that probes, which are one of most important users of this, are not specified in topology at all, and uapi for them was already specified and passed down form cAVS.
Also, I don't see that alleviating the hassle of binding multiple consumers to single message. How do you see that part solved?

lgirdwood replied:
I would class probes as a system debug user with direct access to buffer (including methods in the buffer structures). IPC will use the buffer ID and bind it to a SW probe.
Are there any other use cases for multiple consumers to a single user (aside from our current system uses today).

jajanusz replied:
@lgirdwood I don't understand why you want to have 1-1 relationship for events. For me it's counter-intuitive. I'd bet it's counter-intuitive for most of people as I usually see event notifiers to be always implemented as 1-to-many in various SDKs (when we consider one central notifier). What you want to force here is some kind of hard bind, but it shouldn't be named notifier then, but something more like observer+observable.

mmaka1 commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,163 @@
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX)
```
Note there are static code analyzers that interpret enum as signed type and require to add also < <min_value> check.

mmaka1 commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,163 @@
+{
+	int i;
+
+	for (i = 0; i < NOTIFIER_ID_MAX; i++)
```
Why do not init with the first symbol from enum list?

akloniex replied:
done

mmaka1 commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,163 @@
 	*notify = rzalloc(RZONE_SYS, SOF_MEM_CAPS_RAM, sizeof(**notify));
 
-	list_init(&(*notify)->list);
+	for (i = 0; i < NOTIFIER_ID_MAX; i++)
```
The comment about initialization applies here as well.

mmaka1 commented on src/lib/notifier.c:
```diff
@@ -37,13 +44,19 @@ int notifier_register(void *receiver, void *caller, enum notify_id type,
 
-	if (type >= NOTIFIER_ID_MAX)
+	if (type >= NOTIFIER_ID_MAX) {
+		trace_notifier_error("notifier_register() error: invalid ID to "
```
Would be useful to print out the invalid id.

mmaka1 commented on src/lib/notifier.c:
```diff
@@ -64,8 +77,11 @@ void notifier_unregister(void *receiver, void *caller, enum notify_id type)
 
-	if (type >= NOTIFIER_ID_MAX)
+	if (type >= NOTIFIER_ID_MAX) {
+		trace_notifier_error("notifier_unregister() error: invalid ID "
```
Print out the detected invalid ID.

akloniex commented on src/audio/kpb.c:
```diff
@@ -408,13 +407,13 @@ static int kpb_prepare(struct comp_dev *dev)
-	/* Register KPB for async notification */
-	notifier_register(&kpb->kpb_events);
+	/* Register KPB for notification */
+	if (!notifier_register(dev, NULL, NOTIFIER_ID_KPB_CLIENT_EVT,
```
That's true, good catch.

akloniex commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,163 @@
+		if (core_mask & NOTIFIER_TARGET_CORE_MASK(i)) {
 			if (i == cpu_get_id()) {
-				notifier_notify();
+				notifier_notify(caller, type, data);
```
@mmaka1 I'm not sure if that's a good approach. To me it would look much better to have a wrapper with field specifying amount of elements in array, if you ever needed to pass an array via callback in the first place. Also, it would force all receivers to match signature with `data_size`, which would be unused in vast majority of cases.
If you insist I can add that, but I don't like it.

akloniex commented on src/lib/notifier.c:
```diff
@@ -14,82 +14,163 @@
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX)
```
Changed to assertion

lyakh commented on src/include/sof/lib/notifier.h:
```diff
@@ -19,42 +19,31 @@ struct sof;
+	NOTIFIER_ID_SSP_FREQ,			/* struct clock_notify_data * */
+	NOTIFIER_ID_KPB_CLIENT_EVT,		/* struct kpb_event_data * */
+	NOTIFIER_ID_DMA_DOMAIN_CHANGE,		/* struct dma_chan_data * */
+	NOTIFIER_ID_MAX
```
It looks like `ID_MAX` is the first invalid ID, so, it isn't the maximum (valid) ID any more, maybe call it `ID_NUMBER` or `ID_COUNT`? That's also how you use that value, it seems.

akloniex replied:
done

akloniex commented:
@lbetlej 1 warning and 1 check cleared out, 2 checks remaining, regarding alignment, which I'm unwilling to change, as it would introduce a lot of mess (callback signature doesn't fit into line continuation, if aligned with parenthesis).

lyakh commented on src/drivers/imx/edma.c:
```diff
@@ -195,10 +198,14 @@ static int edma_stop(struct dma_chan_data *channel)
-	struct dma_cb_data next = { .elem.size = bytes };
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes
```
not a reason for a new version on its own, but if you do have to make a new version, maybe you can also add a comma above after `bytes` (and in similar places, that your patches touch, if any). As we have seen recently it makes potential future patching easier.

akloniex replied:
done

lyakh commented on test/cmocka/src/notifier_mocks.c:
```diff
@@ -0,0 +1,102 @@
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
```
no need for "safe"

akloniex replied:
It's a mock, nobody really cares, and this also reflects how original implementation looks like.

lyakh commented on src/audio/buffer.c:
```diff
@@ -174,6 +180,11 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes)
+	struct buffer_cb_transact cb_data = {
+		.buffer = buffer,
+		.transaction_amount = bytes,
+		.transaction_begin_address = buffer->r_ptr
```
as far as I now understand, this PR will have updates, so you'll have a chance to add commas to struct initialisations like this one

akloniex replied:
done

lyakh commented on src/audio/dai.c:
```diff
@@ -540,6 +542,11 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
+	/* setup callback */
+	notifier_unregister(dev, dd->chan, NOTIFIER_ID_DMA_COPY);
+	notifier_register(dev, dd->chan, NOTIFIER_ID_DMA_COPY,
+			  dai_dma_cb);
```
is this because there's no suitable initialisation location, where a notifier can be registered once?

akloniex replied:
True. Previously these callbacks were registered in config/params, but in multicore scenario, trigger is the first call done on target core, so to register to appropriate notifier (there is 1 per core), I've moved it here.
@tlauda is currently working on moving whole pipeline setup to target core, instead of always building them on master core, which would allow us to put registration back to config/params calls.

akloniex commented on src/lib/notifier.c:
```diff
@@ -14,82 +15,161 @@
+	/* iterate through notifiers and send event to
+	 * interested clients
+	 */
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
```
Consumer might unsubscribe from within callback handler, e.g. when BUFFER_FREE is called.

akloniex commented on src/lib/notifier.c:
```diff
@@ -14,82 +15,161 @@
+	if (!handle)
+		return -ENOMEM;
+
+	list_init(&handle->list);
```
done

lyakh commented on src/include/sof/lib/notifier.h:
```diff
@@ -54,11 +40,14 @@ struct notifier {
+int notifier_register(void *receiver, void *caller, enum notify_id type,
+		void (*cb)(void *arg, enum notify_id type, void *data));
+void notifier_unregister(void *receiver, void *caller, enum notify_id type);
+void notifier_unregister_all(void *receiver, void *caller);
```
I think a few of those pointers in the above 3 declarations can be made `const`

lyakh commented on src/include/sof/lib/notifier.h:
```diff
@@ -54,11 +40,14 @@ struct notifier {
-void notifier_event(struct notify_data *notify_data);
+void notifier_notify_remote(void);
+void notifier_event(void *caller, enum notify_id type, uint32_t core_mask,
+		    void *data, uint32_t data_size);
```
`const void *caller` ?

lyakh commented on src/lib/notifier.c:
```diff
@@ -14,82 +15,160 @@
-
-void notifier_register(struct notifier *notifier)
+struct notify_data {
+	void *caller;
```
`const` ?

lyakh commented on src/lib/notifier.c:
```diff
@@ -14,82 +15,160 @@
+struct callback_handle {
+	void *receiver;
+	void *caller;
+	void (*cb)(void *arg, enum notify_id, void *data);
```
all or some of 3 `const` ?

jajanusz commented:
codecheck down again here, but I checked that it's actually ok

lyakh commented on src/lib/notifier.c:
```diff
@@ -14,82 +15,160 @@
+
+	handle = rzalloc(RZONE_SYS_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*handle));
+
+	if (!handle)
```
It's common to put allocation and NULL verification directly in 2 subsequent lines, with no empty line inbetween.

lyakh commented on test/cmocka/src/notifier_mocks.c:
```diff
@@ -0,0 +1,102 @@
+
+	handle = rzalloc(0, 0, sizeof(struct callback_handle));
+
+	if (!handle)
```
ditto - empty line

jajanusz commented:
@lyakh You can review even after merge so author or someone else will consider solving these issues in next PRs. I know it would be best to merge perfect PR, but sometimes when PR is stalled so long you have to go for **good enough**, because costs of depending features are snowballing. So as long as it's cosmetic and doesn't create real technical debt we are ok to solve it later with no additional cost.
